### ISSUE TITLE
High CPU Usage by System Extension even when main app is not open

### ISSUE THREAD

**Issue by ideologysec** [SuprHackerSteve/Crescendo#13]
During a round of App Store updates today, the installd process pegged a cpu core (expected, sadly), but the com.superhackersteve extension also started taking up ~40% of one core (unexpected!). I assume this is due to a high volume of installd/EndpointSecurity events being generated; it was certainly surprising when the app wasn't open, however.

Is there any way to unload the extension when the app isn't open? Thanks!

**Comment by SuprHackerSteve** [comment]
@ideologysec I am still exploring ESF and how this is supposed to work efficiently. It doesn't look like apple wants to unload and load the extension with each use. The fact of the matter is if you have high IO there will always be high load on monitoring these events.

**Comment by ideologysec** [comment]
I definitely expect high overhead from monitoring high I/O operations, just not while the app wasn't open (the difference between "active monitoring" and "continuous monitoring", I suppose.)

If it's a limitation of the System Extension/ESF, I understand. Just thought I'd raise the issue in case there was something to be done.

**Comment by SuprHackerSteve** [comment]
@ideologysec Appreciate the feedback. I'll dig in and let you know what I come up with as a solution.

**Comment by SuprHackerSteve** [comment]
@ideologysec I did some investigating. I only see 2-3% CPU utilization when the app is not running even at exercising heavy process events and file system IO in a tight loop on a modern Macbook Pro. There is not much processing going on in the system extension when a client app is not connected. Could you confirm you are seeing this in the system extension? What hardware are you using to measure performance?

Either way, I dug into this and I came up with a solution. The system extension will simply timeout after a disconnection over a time period and dispose of the `esclient`. I'm going to mull this over for a few days.

**Comment by ideologysec** [comment]
I'm running a 15" 2019 MacBook Pro, Catalina 10.15.4 (non-Supplemental Update).

Reason I filed the issue:
The round of iWork updates last week caused installd to peg at least one core; while that was happening the main app was not open (hadn't been in days) but the extension was showing 30-40%

Follow Up:
Today I did an incremental clone in CarbonCopyCloner to an external SSD, and again saw the extension using 40-60% CPU - though it jumps around a bit.
<img width="280" alt="image" src="https://user-images.githubusercontent.com/790061/78941681-fee99900-7a6c-11ea-8b0a-e6e2be89a5a9.png">
<img width="280" alt="image" src="https://user-images.githubusercontent.com/790061/78941698-0446e380-7a6d-11ea-97f9-29c8fa4e547f.png">

I'm happy to do additional troubleshooting, provide configurations, etc - just let me know what would be helpful.
